### PR TITLE
Hn/sft recipe bugfix

### PR DESCRIPTION
**Pull request by hieuchi911**
# Description

The current SFT recipe `ForgeSFTRecipe` does not apply gradient clipping, causing the training loss to start increasing after some point due to unstable training. Plus, `seq_len` is not being applied during tokenization.

## Test Plan

Comparing training runs with and without gradient clipping, the run with gradient clipping shows a stabilized training loss curve, while the run without gradient clipping shows increasing or unstable loss: https://api.wandb.ai/links/p1ai/qrnczlpx

### CODE REVIEW

File: src/forge/data/tokenizer.py [hunk 1]
```diff
@@ -229,6 +229,7 @@ def __init__(
         generation_config_path: str | None = None,
         chat_template_path: str | None = None,
         truncation_type: str = "right",
+        max_seq_len: int | None = None,
```
Comment (felipemello1) [inline src/forge/data/tokenizer.py:233]: please add to the docstring

File: src/forge/data/tokenizer.py [hunk 2]
```diff
@@ -259,6 +260,8 @@ def __init__(
         for token in self.special_tokens:
             self.special_tokens_mapping[token] = self.base_tokenizer.encode(token)
 
+        self.max_seq_len = max_seq_len
```
Comment (felipemello1) [inline src/forge/data/tokenizer.py]: lets put this at the top of the function

File: src/forge/data/tokenizer.py [hunk 3]
```diff
@@ -294,6 +297,10 @@ def tokenize_messages(
         add_eos: bool = True,
         max_seq_len: int | None = None,
     ) -> tuple[list[int], list[bool]]:
+
+        if max_seq_len is None and self.max_seq_len is not None:
+            max_seq_len = self.max_seq_len
```
Comment (felipemello1) [inline src/forge/data/tokenizer.py]: since the function takes max_seq_len as an arg, lets remove this check and pass the max_seq_len in the _call_ instead

File: src/forge/data/tokenizer.py
Comment (felipemello1) [inline src/forge/data/tokenizer.py]:
```suggestion
        tokens, mask = self.tokenize_messages(messages, max_seq_len=self.max_seq_len)  # TODO: add_end_tokens
```

File: apps/sft/main.py [hunk 4]
```diff
@@ -281,9 +287,20 @@ def train_step(self, batch) -> None:
...
 
+        grad_norm = dist_utils.clip_grad_norm_(
+            [p for m in self.model_parts for p in m.parameters()],
+            self.job_config.training.max_norm,
+            foreach=True,
+            pp_mesh=(
+                parallel_dims.world_mesh["pp"] if parallel_dims.pp_enabled else None
+            ),
+            ep_enabled=parallel_dims.ep_enabled,
+        )
```
Comment (felipemello1) [inline apps/sft/main.py:298]:
```suggestion
        grad_norm = dist_utils.clip_grad_norm_(
            [p for m in self.model_parts for p in m.parameters()],
            self.job_config.training.max_norm,
            foreach=True,
            pp_mesh=parallel_dims.get_optional_mesh("pp"),
            ep_enabled=parallel_dims.ep_enabled,
        )
```

File: apps/sft/main.py [hunk 5]
```diff
@@ -114,6 +114,11 @@ async def setup(self):
         # Load training datasets
         logger.info("Setting training datasets")
         train_datasets_config = self.job_config.training.datasets
+        if self.job_config.training.seq_len is not None:
+            self.max_seq_len = self.job_config.training.seq_len
+        else:
+            self.max_seq_len = None
```
Comment (felipemello1) [inline apps/sft/main.py]: i think that we can delete this and use `self.job_config.training.seq_len` directly. Wdyt?

File: apps/sft/main.py (hunk 4 above)
Comment (hieuchi911) [reply apps/sft/main.py:298]: Hi @felipemello1 I found that `parallel_dims.get_optional_mesh` is not available in `torchtitan==0.2.0` but `torchtitan==0.2.1`, and `torchtitan==0.2.1` requires `torch==2.11.0`? Either way I'm getting dependencies conflict when trying to install these versions

File: apps/sft/main.py (hunk 5 above)
Comment (hieuchi911) [reply apps/sft/main.py]: That makes sense, this extra `self.max_seq_len` is not necessary, it's more of a tokenizer thing so having that in `HuggingFaceModelTokenizer` should be enough, I removed it and everything works fine

File: apps/sft/main.py (hunk 4 above)
Comment (felipemello1) [reply apps/sft/main.py:298]: sry, i was not aware! we are updating the torchtitan version, but we can keep as it was for now. my bad

**Comment by felipemello1** [comment]
can you please do the meta CLA check so i can merge when ready? will finish reviewing it on monday. Ty for the changes!
